### DISCUSSION TITLE
URL-safe `input` compression.

### DISCUSSION THREAD

**Discussion by jlalmes** [trpc/trpc#1936] [Ideas]
💡 Idea

## URL-safe `input` compression.

When sending queries with huge inputs - it is likely you will hit URL character limits. Should we consider adding a URL-safe compression to the input query string?

Think of this as a more space-efficient `encodeURIComponent(JSON.stringify(input))`.

```curl
# before
GET /bigInput?input=%7B%22ids%22%3A%5B%22bd190014-40af-4a27-a74d-a756bea23eb9%22%22...
# after
GET /bigInput?compressed=1&input=XQAAgADIAQAAAAAAAAA9iIkmgZNThWHvkoxYOhkMap8xuPc5WJ...
```

### Compression algos: 

* lz-string: https://github.com/pieroxy/lz-string
* JSONCrush - https://github.com/KilledByAPixel/JSONCrush
* rison - https://github.com/Nanonid/rison
* json-url - https://github.com/masotime/json-url 

### Considerations
* What is the time/efficiency trade-off? (Benchmarks required).
* Should we only apply compression if we know the input payload is going to surpass some `maxURLLength`?
* Should this be an opt-in transformer only (like `superjson`)?

Feedback and thoughts?

**Comment by KATT** [comment]
If we go down this route what we're really talking about is straying away from JSON as the defacto mode of transport. If we do this we should also have support for some sort of binary format when doing `POST`s that could take the compression further and let us have something else than `content-type: application/json`.

Furthermore, I think the server should be able to receive different types of request types and be informed on what to do. I have thought about this a bit before in the context of transformers, so outlining below.

## Some ideas for communicating transformers/serializer

> Rough ideas below, I think best by writing code, but don't see any of this as a final design.

### `contentType`-param

Goals:

- Allow for multiple content-types
- Allow for deserializing from a `Response` with a `body`
- Allow for serializing into a `Fetch.body`
- Allow for serializing into a string for query params

**When creating a server:**

```ts
import { initTRPC } from '@trpc/server';
import { jsonContentType, createContentType } from '@trpc/server/content-type';

// JSON crush could be a community plugin that looks something like this:
import jsoncrush from 'jsoncrush';
const crush = createContentType({
  key: 'crush',
  contentType: 'application/json',
  // a smarter content type would use something that could be streamed, i.e ReadableStream from `Response`
  // used for `POST`:
  fromResponse: async (res) => jsoncrush.uncrush(await res.text()),
  fromString: (str) => jsoncrush.uncrush(str),
  toBody: data => jsoncrush.crush(data), // this could allow for the types that the `Fetch API` allows - ArrayBuffer, string, [...]
  // used for `GET`
  toString: data => jsoncrush.data(data),
})

const trpc = initTRPC({
  contentTypes: [
    jsonContentType,
    crush,
  }
});
```

**HTTP level**

```
// inform server that we are sending `crush` content
GET http://localhost?content=crush&input=[...]
GET http://localhost?content=json&input=[...]

POST http://localhost?content=crush
content-type: "application/json"
// [..]
```

### Similarly, `transformer`-param could be used

```ts
import { initTRPC } from '@trpc/server';
import { defaultTransform } from '@trpc/server/transform';
// JSON crush could be a community plugin
import { crush } from 'trpc-json-crush';

const trpc = initTRPC({
  transforms: [
      defaultTransform, // our "no-op" trnasform
      { key: "superjson", value: superjson },
   ],
});
```

### Bonus, allow for method overriding

I made [this tweet](https://twitter.com/alexdotjs/status/1528854148011130880) yesterday and the main feedback that came back was to be able to sometimes have massive bodies when doing queries.

```
// interpret as query, even if it's a post
POST http://localhost?type=query
```

### The client story

Maybe only needs to allow for one content type.

```ts
import { crush } from "~/shared/contentType"

const client = createTRPCClient({
  links: [ /* ... */ ],
  contentType: crush,
});

```

**Comment by jlalmes** [comment]
From our discussions on [Discord](https://discord.com/channels/867764511159091230/934201568402751558/978305984445435904) it seems like there a number different ideas bouncing around here.

1. URL-safe `input` compression.
    * Pros: Shorter URLs for `queries`.
2. Transformers to replace `JSON.stringify/parse` & `encode/decodeURIComponent`.
    * Pros: Enables full control of payload serialization, cheap implementation (_compared to idea 3_).
3. Support for additional `Content-Type`s (cc. https://github.com/trpc/trpc/issues/1401).
    * Pros: Same as above, but enables use of streams 🔥.
4. Support for procedure method overriding `query`->`mutation`.
    * Pros: Avoid long URLs issue altogether.
    
A combo of 2 & 4 would solve most use-cases & feel like _easier_ wins. I'd love to open idea 3 as it's own issue for future release.

**Comment by omar-dulaimi** [comment]
I'm excited where this discussion is going. I see that tRPC is going back to it's roots, to gRPC. Where requests go as binary, and the payloads are extremely small.

**Comment by jlalmes** [comment]
Have opened the following issues:

* https://github.com/trpc/trpc/issues/1939
* https://github.com/trpc/trpc/issues/1938
* https://github.com/trpc/trpc/issues/1937

**Reply by KATT** [reply]
Awesome. 

I really think #1937 will be the killer feature that can remove the need for transformers.

**Reply by sachinraja** [reply]
#1937 would be amazing. Is that free for anyone to take or did one of you want to work on that?

**Reply by jlalmes** [reply]
Happy for you to pick it up @sachinraja!

**Reply by jlalmes** [reply]
I will take https://github.com/trpc/trpc/issues/1939, and I assume that we can ignore https://github.com/trpc/trpc/issues/1938 if we do https://github.com/trpc/trpc/issues/1937 instead?

**Reply by sachinraja** [reply]
Yes, #1937 will likely replace the current transformers API.

**Comment by Lilja** [comment]
Managed to find [this project some time ago](https://github.com/RainwayApp/bebop). Might be interesting?

**Reply by jlalmes** [reply]
Thanks for sharing @Lilja, but it appears that `bebop` has the same issues as `protobuf`, they achieve type-safety (and hence highly efficient payloads) via codegen, which does not make it suitable for tRPC. _Fully type-safe APIs, without schemas or code generation_ is what differentiates tRPC from other RPC solutions, such as gRPC (cc @omar-dulaimi).

**Reply by sachinraja** [reply]
Additionally, in some preliminary tests I did, switching serialization formats won't help much if your data is mostly unique. What really helps is compression.
